pyo3-polars/src/gil_once_cell.rs

```diff
@@ -0,0 +1,34 @@
+use std::cell::UnsafeCell;
+
+use pyo3::{PyResult, Python};
+
+// Adapted from PYO3 with the only change that
+// we allow mutable access with when the GIL is held
+
+pub struct GILOnceCell<T>(UnsafeCell<Option<T>>);
+
+// T: Send is needed for Sync because the thread which drops the GILOnceCell can be different
+// to the thread which fills it.
+unsafe impl<T: Send + Sync> Sync for GILOnceCell<T> {}
+unsafe impl<T: Send> Send for GILOnceCell<T> {}
+
+impl<T> GILOnceCell<T> {
```

Comment:
This implementation of the `GILOnceCell` can only be used if the users of this library also includes a function that is called "on startup", that initializes the value in the Cell, similar to https://github.com/pola-rs/polars/blob/97eff077f37209386836361e9ab4da582bc5b18e/py-polars/src/on_startup.rs#L98

I don't think that this requirement would be a good idea. We could check in the `with_gil` function if the value in the cell has been initialized, and if not initialize it there